File: Array/23-merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    def mergeList(self, list1, list2):
        dummy = ListNode()

        curr = dummy

        while list1 or list2:
            n1 = list1.val if list1 else float('inf')
            n2 = list2.val if list2 else float('inf')
            if n1<= n2:
                curr.next = list1
                list1 = list1.next
            else:
                curr.next = list2
                list2 = list2.next
            curr = curr.next
        return dummy.next

    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        
        while len(lists) > 2:
            mid = len(lists)//2
            temp = []
            temp.append(self.mergeKLists(lists[:mid+1]))
            temp.append(self.mergeKLists(lists[mid+1:]))
            lists = temp
        
        if len(lists) == 2:
            return self.mergeList(lists[0], lists[1])
        elif len(lists) == 1:
            return lists[0]
        return None

File: JavaScript/23-merge-k-sorted-lists/merge-k-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */

function mergeLists(list1, list2) {


    let dummy = new ListNode(0);
    let tail = dummy;

    while(list1 && list2){
        if(list1 && list2) {
            if(list1.val < list2.val) {
                tail.next = list1;
                list1 = list1.next;
            } else {
                tail.next =  list2;
                list2 = list2.next;
            }
        }
        tail = tail.next;
    }
    tail.next = list1 || list2;
    return dummy.next;
}
var mergeKLists = function(lists) {
    if(!lists || lists.length == 0) return null;

    function helper(left, right) {
        if (left > right) return null;
        if(left == right) return lists[left];
        let mid = Math.floor((left+right)/2);
        let l1 = helper(left, mid);
        let l2 = helper(mid+1, right);
        return mergeLists(l1, l2)
        
    }
    // while(lists.length > 1) {
    //     if(lists.length == 2) {
    //         lists = [mergeLists(lists[0], lists[1])];
    //     } else {
    //         let mid = Math.floor(lists.length/2);
    //         lists = [mergeKLists(lists.slice(0, mid)), mergeKLists(lists.slice(mid, lists.length))];

    //     }

    // }

    return helper(0, lists.length-1);
};